akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2.scala

```diff
@@ -190,29 +190,40 @@ private[http] final class Http2Ext(private val config: Config)(implicit val syst
 
   def outgoingConnection(
     host:              String,
-    port:              Int                      = 443,
+    port:              Int                      = -1,
     settings:          ClientConnectionSettings = ClientConnectionSettings(system),
-    connectionContext: HttpsConnectionContext   = Http().defaultClientHttpsContext,
+    connectionContext: ConnectionContext        = Http().defaultClientHttpsContext,
     log:               LoggingAdapter           = system.log): Flow[HttpRequest, HttpResponse, Any] = {
-    def createEngine(): SSLEngine = {
-      val engine = connectionContext.sslContextData match {
-        // TODO FIXME configure hostname verification for this case
-        case Left(ssl) =>
-          val e = ssl.sslContext.createSSLEngine(host, port)
-          TlsUtils.applySessionParameters(e, ssl.firstSession)
-          e
-        case Right(e) => e(Some((host, port)))
-      }
-      engine.setUseClientMode(true)
-      Http2AlpnSupport.clientSetApplicationProtocols(engine, Array("h2"))
-      engine
+    val (realPort, tls) = connectionContext match {
+      case https: HttpsConnectionContext =>
+        val realPort = if (port == -1) 443 else port
+        (
+          realPort,
+          TLS(() => createEngine(host, realPort, https), closing = TLSClosing.eagerClose)
+        )
+      case other =>
+        (if (port == -1) 80 else port, TLSPlacebo())
     }
 
     Http2Blueprint.clientStack(settings, log) atop
       Http2Blueprint.unwrapTls atop
       LogByteStringTools.logTLSBidiBySetting("client-plain-text", settings.logUnencryptedNetworkBytes) atop
-      TLS(createEngine _, closing = TLSClosing.eagerClose) join
-      settings.transport.connectTo(host, port, settings)
+      tls join
+      settings.transport.connectTo(host, realPort, settings)
+  }
+
+  private def createEngine(host: String, port: Int, connectionContext: HttpsConnectionContext): SSLEngine = {
+    val engine = connectionContext.sslContextData match {
+      // TODO FIXME configure hostname verification for this case
+      case Left(ssl) =>
+        val e = ssl.sslContext.createSSLEngine(host, port)
+        TlsUtils.applySessionParameters(e, ssl.firstSession)
+        e
+      case Right(e) => e(Some((host, port)))
+    }
+    engine.setUseClientMode(true)
+    Http2AlpnSupport.clientSetApplicationProtocols(engine, Array("h2"))
+    engine
```

Comment:
Since this is gone in #3511 is there something from this PR to incorporate there rather than merge this?

Reply:
I'd rather keep all the stream setup together in this file, in fact.

Reply:
Let me rephrase that now that it is moved back into `Http2` in that PR, is there something for prior knowledge or regular HTTP/2 in here that I have missed in the connection methods over there. I added a test in `WithPriorKnowledgeSpec` which seems to pass, and so does the sample applications using regular HTTP2 against akka.io, but maybe I missed something.

Reply:
I think  #3511 indeed makes this one redundant. Since that one is so close now I think we can focus on that and close this one.